text/0041-extensible-block-body.md

```diff
@@ -0,0 +1,104 @@
+
+By employing Merkle proofs, we will enable partial verification of block bodies. For example, verifying the list of deploys will require downloading only that list and the hash of the rest of the body, significantly reducing the bandwidth usage. We will also be able to store parts of the body in a way that allows for both easy verification and easy extension of the block contents.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The hash of the block body would be redefined as the following:
+
+- hash of the body is the hash of `[hash of the deploy hashes list, hash_2]`
+- `hash_2` is the hash of `[hash of the transfer hashes list, hash_3]`
+- `hash_3` is the hash of `[hash of the proposer, hash_4]`
+- `hash_4` is the hash of an empty bytestring.
+
+This can be illustrated by the following diagram:
+
+```
+                   body_hash
+                   /      \
+ hash(deploy_hashes)      hash_2____________
+                         /                  \
+                   hash(transfer_hashes)     hash_3_______
+                                            /             \
+                                         hash(proposer)   hash("")
+```
+
+So, for example, to securely download only the transfer hashes stored in a block with a known body hash, we will additionaly require only the hash of the deploy hashes list and `hash_3`. This will allow us to calculate `hash_2` and `body_hash`, and to verify that the body hash matches the one we have.
+
+The block bodies would be stored the following way:
+- A `block_body_merkle` database would store a key-value map of `body_hash → vec![hash(deploy_hashes), hash(transfer_hashes), hash(proposer)]`. The type `Vec<Digest>` of the value will allow for storing blocks with more fields in the same database without the risk of causing problems with deserialization.
+- A `deploy_hashes` database would store the lists of deploy hashes keyed by `hash(deploy_hashes)`.
+- A `transfer_hashes` database would store the lists of transfer hashes keyed by `hash(transfer_hashes)`.
+- A `proposer` database would store the proposers keyed by `hash(proposer)`.
+
+The new way of hashing and new databases would start being used in a specific, chosen protocol version. The blocks in this protocol version and above would have their bodies hashed using the new way and be stored in the new databases. Older blocks would be kept in the old database.
+
+This is to ensure that we can still verify the integrity of storage after we transition to the new format. Changing the body hash of the old blocks would also change the hash of the block and hashes of all the later blocks, so it is not an option. So in order to be able to verify that the stored data still corresponds to the hashes from the blocks, we need to be able to calculate them the old way even after we start using the new way for new blocks.
```

Comment:
Maybe the integrity verification should be version-based and we would have different pieces of logic for different versions: on node startup, during the storage integrity verification phase, we will iterate over the blocks in the storage, load each into memory, check the protocol version and verify the integrity with appropriate logic. This doesn't require changing the storage at all.